Expose a failure reason in `PaymentFailed`

alecchendev opened the pull request:
Closes #2095. Adds a new public enum `PaymentFailureReason` and a new field on `Event::PaymentFailed` to communicate why the payment failed completely.

alecchendev commented on lightning/src/events/mod.rs:
```diff
@@ -270,6 +270,41 @@ impl_writeable_tlv_based_enum!(InterceptNextHop,
+	UserAbandoned,
+	/// We exhausted all of our retry attempts while trying to send the payment, or we
+	/// exhausted the `Retry::Timeout` if the user set one.
+	ExhaustedRetryAttempts,
+	/// The payment expired while retrying. If the user chooses to include an `expiry_time` in their
+	/// `PaymentParams` this payment will be failed if it is retried after this time has elapsed.
+	PaymentExpired,
+	/// We failed to find a route while retrying the payment.
+	FailedRoutingRetry,
+	/// The amount we would pay if we were to retry the payment would exceed the intended amount
+	/// above our allowed threshold.
+	RetryAmountOverflow,
+	/// Parameter error while sending with retry.
+	RetryParameterError,
```
I’m thinking should I consolidate these different retry errors (or group them under their own enum?)? They communicate more information, but I’m not really sure how practically helpful they are to a user.

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1176,6 +1178,7 @@ impl OutboundPayments {
 						full_failure_ev = Some(events::Event::PaymentFailed {
 							payment_id: *payment_id,
 							payment_hash: payment.get().payment_hash().expect("PendingOutboundPayments::RetriesExceeded always has a payment hash set"),
+							reason: PaymentFailureReason::RecipientRejected
```
For `PaymentFailureReason::RecipientRejected` here, it seems difficult to distinguish whether it’s actually because the recipient rejected the payment or because the payment failed somewhere along the path (will probably change the name to reflect this), not sure if this is worth trying to do or if it can be easily done since from my understanding of onion error messages we’d probably have to look at the actual error data?

alecchendev replied:
Oh nice!

alecchendev replied:
Added this, along with `FailedAlongPath` variant, and used similar conditions (relating to failing permanently or being retryable) in tests to accomodate for this change

alecchendev reviewed (commented):
Some things I’m considering changing right now if anyone wants to give some feedback:

alecchendev commented on lightning/src/events/mod.rs:
```diff
@@ -1158,13 +1196,16 @@ impl MaybeReadable for Event {
+					let mut reason = UpgradableRequired(None);
 					read_tlv_fields!(reader, {
 						(0, payment_id, required),
+						(1, reason, upgradable_required),
```
Oops, thanks

alecchendev commented:
Squashed to pass CI and because it was getting messy to read all the fixups since I was changing/merging enum variants and had to undo work from previous commits.

alecchendev commented on lightning/src/events/mod.rs:
```diff
@@ -270,6 +270,41 @@ impl_writeable_tlv_based_enum!(InterceptNextHop,
+	FailedRoutingRetry,
+	/// The amount we would pay if we were to retry the payment would exceed the intended amount
+	/// above our allowed threshold.
+	RetryAmountOverflow,
```
Merged into `UnexpectedError` variant

valentinewallace commented on lightning/src/events/mod.rs:
```diff
@@ -270,6 +270,41 @@ impl_writeable_tlv_based_enum!(InterceptNextHop,
+pub enum PaymentFailureReason {
+	/// The intended recipient rejected our payment.
+	RecipientRejected,
+	/// The payment failed while being forwarded along the path. A [`Event::PaymentPathFailed`]
+	/// should have come before this.
+	FailedAlongPath,
```
I wonder if we should consolidate this with `ExhaustedRetryAttempts`, because in the 1 place we return this error, I think retries being exhausted is arguably the "real" reason for the `PaymentFailed` event, and the `FailedAlongPath` part is communicated via the `PaymentPathFailed` events. cc @TheBlueMatt does that make sense?

alecchendev replied:
> Sadly doing this is going to require storing the reason in the `Abandoned` payment case - we don't know without context whether we were `UserAbandoned` or `ExhaustedRetryAttempts`.

Just to clarify did you mean `RecipientRejected` here instead of `UserAbandoned`?

As I was looking into trying to store a reason in `Abandoned` it seemed like we didn't really have what we needed to store the correct reason for what was happening, which has led me to a question. I'm sort of confused on what actually would differentiate a situation where a recipient rejected our payment versus our path failed and we exhausted our last retry. When a recipient rejects our payment, we also get a path failure don't we (just with different onion error data)?

alecchendev replied:
Hmm, I'm confused about how we can't tell between user abandoned/exhausted. I'm thinking that the only time that `UserAbandoned` is used is in `ChannelManager::abandon_payment` which just calls `OutboundPayments::abandon_payment` which directly pushes the `PaymentFailed` with that reason?

alecchendev replied:
Yep, sounds good

alecchendev commented:
Separating my fixup commits so they're easier to squash later even though it'll fail CI how they currently are. I am still figuring out the best way to resolve my last commit merging `FailedAlongPath` into `RetriesExhausted`.

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -642,14 +646,14 @@ impl OutboundPayments {
 			if path.len() == 0 {
 				log_error!(logger, "length-0 path in route");
-				self.abandon_payment(payment_id, pending_events);
+				self.abandon_payment(payment_id, PaymentFailureReason::RouteNotFound, pending_events);
```
Oh true!

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1233,15 +1248,16 @@ impl OutboundPayments {
 					pending_events.lock().unwrap().push(events::Event::PaymentFailed {
 						payment_id,
 						payment_hash: payment.get().payment_hash().expect("PendingOutboundPayments::RetriesExceeded always has a payment hash set"),
+						reason: Some(reason),
```
Should this maybe also be the behavior of `PendingOutboundPayment::mark_abandoned`? I.e. should it only fill in the reason if it's not already abandoned (as opposed to right now it just overrides the reason)

alecchendev replied:
I also just pulled from existing state if there is one in the last place where `PaymentFailed` is pushed (in `check_retry_payments`) as it seems like generally if a payment is already abandoned with a certain reason, that's the one we should push

alecchendev replied:
...and one more place in the macro `abandon_with_entry`, even though when this is used the payment is most likely going to be being abandoned for the first time, I figure it should be consistent.

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1176,6 +1186,11 @@ impl OutboundPayments {
 						full_failure_ev = Some(events::Event::PaymentFailed {
 							payment_id: *payment_id,
 							payment_hash: payment.get().payment_hash().expect("PendingOutboundPayments::RetriesExceeded always has a payment hash set"),
+							reason: Some(if payment_retryable {
```
Sounds good, didn't know if we wanted to do that here or in the bug fix, but this'll probably make things easier

valentinewallace reviewed (commented):
Looking good!

valentinewallace commented:
Feel free to squash!

valentinewallace reviewed (commented):
LGTM after squash

alecchendev commented:
Rebased and fixed 2 lines for CI - `payment_tests.rs` line 2924 (the change here doesn't really relate to this PR, don't really know what's normal to do here) and 2959

alecchendev commented:
Rebased again so it's not showing the overlapping change :)

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -803,12 +812,13 @@ impl OutboundPayments {
 				// initial HTLC-Add messages yet.
 			},
 			PaymentSendFailure::PathParameterError(results) => {
+				log_error!(logger, "Failed to send to route due to parameter error in a single path. Your router is buggy");
```
Just added logging the `APIError`(s) in `push_path_failed_evs_and_scids` - I figure in the other cases where we are still able to retry the payment it's still helpful to have the record of those errors

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -824,14 +824,15 @@ impl OutboundPayments {
 		debug_assert_eq!(paths.len(), path_results.len());
 		for (path, path_res) in paths.into_iter().zip(path_results) {
 			if let Err(e) = path_res {
+				log_error!(logger, "Failed to send along path due to error: {:?}", e);
 				if let APIError::MonitorUpdateInProgress = e { continue }
```
Not a big deal, but ideally we shouldn't `log_error` if we get `APIError::MonitorUpdateInProgress`, because it's not an error (🙄). Could conditionally log or move the current log to after the `continue`

alecchendev replied:
ah thanks

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1232,7 +1226,7 @@ impl OutboundPayments {
 					PaymentFailureReason::RetriesExhausted
 				};
-				let _ = payment.get_mut().mark_abandoned(reason); // we'll only Err if it's a legacy payment
+				payment.get_mut().mark_abandoned(reason); // we'll only Err if it's a legacy payment
```
Can remove the comment

valentinewallace reviewed (commented):
LGTM after squash
